src/subscription.rs

```diff
@@ -24,76 +95,74 @@ impl<TReq: Debug + Eq + Hash, TRes: Debug> fmt::Debug for SubscriptionRegistry<T
+            subscription.cancel(true);
+        }
+
+        task::block_on(async { self.subscriptions.lock().await.insert(id, subscription) });
+
+        SubscriptionFuture {
+            id,
+            subscriptions: Arc::clone(&self.subscriptions),
         }
-        SubscriptionFuture { subscription }
     }
 
-    pub fn finish_subscription(&mut self, req: &TReq, res: TRes) {
-        if let Some(subscription) = self.subscriptions.remove(req) {
-            subscription.lock().unwrap().wake(res);
+    pub fn finish_subscription(&self, req: &Request, res: TRes) {
+        let mut subscriptions = task::block_on(async { self.subscriptions.lock().await });
+
+        for sub in subscriptions.values_mut() {
+            if let Subscription::Pending { request, .. } = sub {
+                // wake up all tasks related to the requested resource
+                if request.kind == req.kind {
+                    sub.wake(res.clone());
+                }
+            }
         }
     }
 
     /// After shutdown all SubscriptionFutures will return Err(Cancelled)
-    pub fn shutdown(&mut self) {
-        if self.shutting_down {
+    pub fn shutdown(&self) {
+        if self.shutting_down.swap(true, Ordering::SeqCst) {
             return;
         }
-        self.shutting_down = true;
 
         log::debug!("Shutting down {:?}", self);
 
         let mut cancelled = 0;
-        let mut pending = Vec::new();
-
-        for (_, sub) in self.subscriptions.drain() {
-            {
-                if let Ok(mut sub) = sub.try_lock() {
-                    sub.cancel();
-                    cancelled += 1;
-                    continue;
-                }
-            }
-            pending.push(sub);
-        }
-
-        log::trace!(
-            "Cancelled {} subscriptions and {} are pending (not immediatedly locked)",
-            cancelled,
-            pending.len()
-        );
-
-        let remaining = pending.len();
+        let mut subscriptions = mem::take(&mut *task::block_on(async {
+            self.subscriptions.lock().await
+        }));
 
-        for sub in pending {
-            if let Ok(mut sub) = sub.lock() {
-                sub.cancel();
-            }
+        for (_idx, mut sub) in subscriptions.drain() {
+            sub.cancel(true);
+            cancelled += 1;
         }
 
-        log::debug!(
-            "Remaining {} pending subscriptions cancelled (total of {})",
-            remaining,
-            cancelled + remaining
-        );
+        log::trace!("Cancelled {} subscriptions", cancelled);
     }
 }
 
-impl<TReq: Debug + Eq + Hash, TRes: Debug> Default for SubscriptionRegistry<TReq, TRes> {
+impl<TRes: Debug + Clone + PartialEq> Default for SubscriptionRegistry<TRes> {
     fn default() -> Self {
         Self {
             subscriptions: Default::default(),
-            shutting_down: false,
+            shutting_down: Default::default(),
         }
     }
 }
```

Comment:
Cannot this be `#[derive(Default)]` because it adds some `TRes: Default` bounds or is there some other issue I cannot see?

Reply:
yep, and `TRes` unfortunately can't be `Default`